src/servers/grpc_server.cc

```diff
@@ -249,46 +252,75 @@ GRPCServer::~GRPCServer()
-      inferenceService->RegisterRPC<StreamInferContext>(
-          &GRPCService::AsyncService::RequestStreamInfer);
-
-  LOG_INFO << "Register Status RPC";
-  (*grpc_server)->rpcStatus_ = inferenceService->RegisterRPC<StatusContext>(
-      &GRPCService::AsyncService::RequestStatus);
-
-  LOG_INFO << "Register Profile RPC";
-  (*grpc_server)->rpcProfile_ = inferenceService->RegisterRPC<ProfileContext>(
-      &GRPCService::AsyncService::RequestProfile);
-
-  LOG_INFO << "Register Health RPC";
-  (*grpc_server)->rpcHealth_ = inferenceService->RegisterRPC<HealthContext>(
-      &GRPCService::AsyncService::RequestHealth);
+  // Group by Port numbers
+  std::map<int32_t, std::vector<std::string>> port_map;
+  size_t i;
+  for (i = 0; i < service_endpoint_count_; i++) {
+    if (endpoint_ports[i] != -1) {
+      port_map[endpoint_ports[i]].push_back(endpoint_names[i]);
+    }
+  }
 
+  // launch on ports
+  if (!port_map.empty()) {
+    i = 0;
+    for (auto const& ep_map : port_map) {
+      g_Resources = std::make_shared<AsyncResources>(
+          server, 1, /* infer threads */ 1 /* mgmt threads */);
+
+      std::string addr = "0.0.0.0:" + std::to_string(ep_map.first);
+      LOG_INFO << "Starting GRPCService at " << addr;
+      (*grpc_servers)[i].reset(new GRPCServer(addr));
+
+      (*grpc_servers)[i]->GetBuilder().SetMaxMessageSize(MAX_GRPC_MESSAGE_SIZE);
+
+      LOG_INFO << "Register TensorRT GRPCService";
+      auto inferenceService =
+          (*grpc_servers)[i]->RegisterAsyncService<GRPCService>();
+
+      for (auto ep_name : ep_map.second) {
+        if (ep_name == "infer") {
+          LOG_INFO << "Register Infer RPC";
+          (*grpc_servers)[i]->rpcInfer_ =
+              inferenceService->RegisterRPC<InferContext>(
+                  &GRPCService::AsyncService::RequestInfer);
+
+          LOG_INFO << "Register StreamInfer RPC";
+          (*grpc_servers)[i]->rpcStreamInfer_ =
+              inferenceService->RegisterRPC<StreamInferContext>(
+                  &GRPCService::AsyncService::RequestStreamInfer);
+        }
+
+        if (ep_name == "status") {
+          LOG_INFO << "Register Status RPC";
+          (*grpc_servers)[i]->rpcStatus_ =
+              inferenceService->RegisterRPC<StatusContext>(
+                  &GRPCService::AsyncService::RequestStatus);
+        }
+
+        if (ep_name == "profile") {
+          LOG_INFO << "Register Profile RPC";
+          (*grpc_servers)[i]->rpcProfile_ =
+              inferenceService->RegisterRPC<ProfileContext>(
+                  &GRPCService::AsyncService::RequestProfile);
+        }
+
+        if (ep_name == "health") {
+          LOG_INFO << "Register Health RPC";
+          (*grpc_servers)[i]->rpcHealth_ =
+              inferenceService->RegisterRPC<HealthContext>(
+                  &GRPCService::AsyncService::RequestHealth);
+        }
+      }
+      i++;
+    }
+  } else {
+    return Status(RequestStatusCode::INVALID_ARG, "must specify valid ports");
```

Comment:
Better error message: "GRPC is enabled but none of the service endpoints have a valid port assignment"